wled00/FX_fcn.cpp

```diff
@@ -837,13 +866,101 @@ uint32_t Segment::getPixelColor(int i)
   }
 #endif
 
+  return strip.getPixelColor(getPixelIndex(i));
+}
+
+uint32_t Segment::getBufferColor(int i) {
+  return strip._buffer[getPixelIndex(i)];
+}
+
+uint16_t Segment::getPixelIndex(int i) {
+  i &= 0xFFFF;
   if (reverse) i = virtualLength() - i - 1;
   i *= groupLength();
   i += start;
   /* offset/phase */
   i += offset;
   if ((i >= stop) && (stop>0)) i -= length(); // avoids negative pixel index (stop = 0 is a possible value)
-  return strip.getPixelColor(i);
+  return i;
+}
+
+void Segment::blendTransition() {
+  switch (transitionStyle) {
+    case TRANSITION_STYLE_PUSH_RIGHT: {
+      uint16_t len = virtualLength();
+      uint16_t pos = (float(progress()) / float(0xFFFFU)) * len;
+      for (int i = 0; i != pos; i++) {
+        setPixelColorExact(i, getPixelColor(len - pos + i));
+      }
+      for (int i = len - pos; i != 0; i--) {
+        setPixelColorExact(i + pos, getBufferColor(i));
+      }
+      break;
+    }
+    case TRANSITION_STYLE_PUSH_LEFT: {
+      uint16_t len = virtualLength();
+      uint16_t pos = (float(progress()) / float(0xFFFFU)) * len;
+      for (int i = len - pos; i != len; i++) {
+        setPixelColorExact(i, getPixelColor(i - len + pos));
+      }
+      for (int i = pos; i != len; i++) {
+        setPixelColorExact(i - pos, getBufferColor(i));
+      }
+      break;
+    }
+    default: {
+      // All other transition styles use the same pixel indexes from the strip and the buffer
+      for (int i = 0; i != virtualLength(); ++i) {
+        uint16_t index = getPixelIndex(i);
+        uint32_t oldCol = strip.getPixelColor(index);
+        uint32_t newCol = strip._buffer[index];
+
+        setPixelColorExact(i, transitionColor(i, oldCol, newCol));
+      }
+      break;
+    }
+  }
+}
+
+uint32_t Segment::transitionColor(int n, uint32_t oldCol, uint32_t newCol) {
```

Comment:
This entire function would need to be incorporated into `currentColor()` to make it transparent.

Reply:
I have reconsidered and this should replace `SEGCOLOR()` macro in some smart way.
`SEGCOLOR()` macro is used throughout effect functions to get currently selected colors (you'd need to implement transitions for all 3 color slots) which ATM `WS2812FX::service()` packs into `strip` global/public variables.

Reply:
It took me a while but I think I understand your point now. What I call "color" here and what is generally known as "color" are different things. Here, it's the final pixel color after the transition style is applied. Generally, "color" is the color slot that the user picks in the UI. Clearly, this method is poorly named. I'll rename it.